vms/platformvm/validators/manager.go

```diff
@@ -164,100 +164,135 @@ func (m *manager) GetValidatorSet(ctx context.Context, height uint64, subnetID i
-	for _, vdr := range currentSubnetValidatorList {
-		primaryVdr, ok := currentPrimaryNetworkValidators.Get(vdr.NodeID)
-		if !ok {
-			// This should never happen
-			return nil, fmt.Errorf("%w: %s", ErrMissingValidator, vdr.NodeID)
-		}
-		vdrSet[vdr.NodeID] = &validators.GetValidatorOutput{
+	currentPrimaryValidatorList := currentPrimaryNetworkValidators.List()
+	primarySet := make(map[ids.NodeID]*validators.GetValidatorOutput, len(currentPrimaryValidatorList))
+	for _, vdr := range currentPrimaryValidatorList {
+		primarySet[vdr.NodeID] = &validators.GetValidatorOutput{
 			NodeID:    vdr.NodeID,
-			PublicKey: primaryVdr.PublicKey,
+			PublicKey: vdr.PublicKey,
 			Weight:    vdr.Weight,
 		}
+
+		// fill PK from primary network
+		if _, found := subnetSet[vdr.NodeID]; found {
+			subnetSet[vdr.NodeID].PublicKey = vdr.PublicKey
+		}
 	}
 
 	for diffHeight := lastAcceptedHeight; diffHeight > height; diffHeight-- {
-		err := m.applyValidatorDiffs(vdrSet, subnetID, diffHeight)
-		if err != nil {
+		if err := m.applyValidatorDiffs(subnetSet, primarySet, subnetID, diffHeight); err != nil {
 			return nil, err
 		}
 	}
 
 	// cache the validator set
-	validatorSetsCache.Put(height, vdrSet)
+	validatorSetsCache.Put(height, subnetSet)
 
 	endTime := m.clk.Time()
 	m.metrics.IncValidatorSetsCreated()
 	m.metrics.AddValidatorSetsDuration(endTime.Sub(startTime))
 	m.metrics.AddValidatorSetsHeightDiff(lastAcceptedHeight - height)
-	return vdrSet, nil
+	return subnetSet, nil
 }
 
 func (m *manager) applyValidatorDiffs(
-	vdrSet map[ids.NodeID]*validators.GetValidatorOutput,
+	targetSet, primarySet map[ids.NodeID]*validators.GetValidatorOutput,
 	subnetID ids.ID,
 	height uint64,
 ) error {
-	weightDiffs, err := m.state.GetValidatorWeightDiffs(height, subnetID)
+	// fully rebuild primary network validators at [height]
+	primaryWeightDiffs, err := m.state.GetValidatorWeightDiffs(height, constants.PlatformChainID)
 	if err != nil {
 		return err
 	}
-
-	for nodeID, weightDiff := range weightDiffs {
-		vdr, ok := vdrSet[nodeID]
-		if !ok {
-			// This node isn't in the current validator set.
-			vdr = &validators.GetValidatorOutput{
-				NodeID: nodeID,
-			}
-			vdrSet[nodeID] = vdr
-		}
-
-		// The weight of this node changed at this block.
-		if weightDiff.Decrease {
-			// The validator's weight was decreased at this block, so in the
-			// prior block it was higher.
-			vdr.Weight, err = math.Add64(vdr.Weight, weightDiff.Amount)
-		} else {
-			// The validator's weight was increased at this block, so in the
-			// prior block it was lower.
-			vdr.Weight, err = math.Sub(vdr.Weight, weightDiff.Amount)
-		}
-		if err != nil {
+	for nodeID, weightDiff := range primaryWeightDiffs {
+		if err := applyWeightDiff(primarySet, nodeID, weightDiff); err != nil {
 			return err
 		}
-
-		if vdr.Weight == 0 {
-			// The validator's weight was 0 before this block so
-			// they weren't in the validator set.
-			delete(vdrSet, nodeID)
-		}
 	}
```

Comment:
Instead of weight diffs, could we use diffValidator? I think the information we require is whether or not this validator added/removed from the set rather than the weight diffs. And weightDiffs could be much bigger than diffValidator.

Reply:
We talked with @abi87, seems what I asked for only applicable for currentStakers which is stored in memory and not related to heights.